src/swarm/neo/util/Batch.d

```diff
@@ -0,0 +1,239 @@
+        *(cast(size_t*)(compress_buf.ptr)) = this.buffer.length;
+
+        // Compress into destination.
+        auto dst = compress_buf[size_t.sizeof .. $];
+        auto compressed_len = lzo.compress(*this.buffer, dst);
+
+        // Minimize dest length and return.
+        compress_buf.length = compressed_len + size_t.sizeof;
+        enableStomping(compress_buf);
+        return compress_buf;
+    }
+
+    public void clear ( )
+    {
+        this.buffer.length = 0;
+        enableStomping(*this.buffer);
+    }
+
+    private size_t sizeOfItem ( T ) ( T item )
+    {
+        static if ( isDynamicArrayType!(T) )
+        {
+            static assert(rankOfArray!(T) == 1,
+                "Only 1d arrays and primitive types are supported, not "
+                ~ T.stringof);
+            return size_t.sizeof + (item.length * ElementTypeOfArray!(T).sizeof);
+        }
+        else
+        {
+            static assert(isPrimitiveType!(T),
+                "Only 1d arrays and primitive types are supported, not "
+                ~ T.stringof);
+            return T.sizeof;
+        }
+    }
+
+    private void addItem ( T ) ( T item )
+    {
+        static if ( isDynamicArrayType!(T) )
+        {
+            static assert(rankOfArray!(T) == 1,
+                "Only 1d arrays and primitive types are supported, not "
+                ~ T.stringof);
+            auto len = item.length;
+            this.addBytes((cast(void*)&len)[0 .. size_t.sizeof]);
+            this.addBytes(item[0..$]);
+        }
+        else
+        {
+            static assert(isPrimitiveType!(T),
+                "Only 1d arrays and primitive types are supported, not "
+                ~ T.stringof);
+            this.addBytes((cast(void*)&item)[0 .. T.sizeof]);
+        }
+    }
+
+    private void addBytes ( in void[] data )
+    {
+        (*this.buffer) ~= data;
+    }
+}
+
+struct BatchReader
+{
+
+}
+
+version ( UnitTest )
+{
+    import ocean.core.Test;
+
+    enum ExpectedResult
+    {
+        Added,
+        TooBig,
+        Sent
+    }
+
+    void checkAdd ( T ) ( cstring test_name, BatchWriter batch, T item,
+        size_t expected_length, ExpectedResult expected = ExpectedResult.Added )
+    {
+        auto t = new NamedTest(test_name);
+        bool sent;
+        auto added = batch.add(item,
+            {
+                sent = true;
+            }
+        );
+
+        with ( ExpectedResult ) final switch ( expected )
+        {
+            case Added:
+                t.test(added);
+                t.test!("==")(batch.get.length, expected_length);
+                t.test(!sent);
+                break;
+
+            case TooBig:
+                t.test(!added);
+                t.test(!sent);
+                break;
+
+            case Sent:
+                t.test(added);
+                t.test!("==")(batch.get.length, expected_length);
+                t.test(sent);
+                break;
+
+            default: assert(false);
```

Comment:
Must be guarded by `version(D_Version2){} else { ... }`

Reply:
Yeah, actually as this is just for a unittest, I'll remove the `final`.